ite: drivers/pinmux: modify pinmux driver

GTLin08 opened the pull request:
Modify the pinmux control method and add support the fun3
& fun4 alternation function.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

jackrosenthal commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,156 @@
+	volatile uint8_t *reg_gpcr =
+		(uint8_t *)(pinmux_config->reg_gpcr + pin);
+
+	if (func == PINMUX_INPUT_ENABLED)
```
nit: zephyr style asks for braces even when not required (unlike linux kernel style)

https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

GTLin08 replied:
Thank you for your review. Braces have been added.

jackrosenthal reviewed (approved):
overall I think this is a good improvement

keith-zephyr commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -42,9 +43,174 @@
+			reg = <0x00f01618 0x0008>;
+			label = "PINMUXB";
+			func3_gcr =     <0xf016f5 0xf016f5 NO_FUNC NO_FUNC
+					 NO_FUNC  NO_FUNC  NO_FUNC 0xf01600>;
+			func3_en_mask = <0x01     0x02     0       0
```
Referencing IT81202 datasheet v0.3.2.1:
For GPIOB0,  and GPIOB1.  The U1CTRL field needs to be set to >0.  Where does this happen?

GTLin08 replied:
U1CTRL>0 has been defined in soc.c(IT8XXX2_GPIO_GRC1 |= BIT(0)).

keith-zephyr replied:
Is this in a different pull request?  I don't see this in soc/riscv/riscv_ite/it82xxx2/soc.c.

keith-zephyr replied:
Never mind, I see it in this PR.

keith-zephyr commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -42,9 +43,174 @@
+			label = "PINMUXE";
+			func3_gcr =     <0xf02032 NO_FUNC  NO_FUNC NO_FUNC
+					 NO_FUNC  0xf016f0 NO_FUNC 0xf02032>;
+			func3_en_mask = <0x01     0        0       0
```
GPIOE0: shows PMER1[1]=1, this sets PMER1[0]=1

GTLin08 replied:
There is a mistake in the specification. PMER1[0]=1 is right.

keith-zephyr commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -42,9 +43,174 @@
+			label = "PINMUXG";
+			func3_gcr =     <0xf016f0 0xf016f0 0xf016f0 NO_FUNC
+					 NO_FUNC  NO_FUNC  0xf016f0 NO_FUNC>;
+			func3_en_mask = <0x20     0x08     0x10     0
```
GPIOG0: SPICTRL[0]>0.  Should the GPIOG0 mask be 0x10?

GTLin08 replied:
There is a mistake in the specification. SPICTRL[1]=1 is right.

keith-zephyr commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -42,9 +43,174 @@
+			func3_gcr =     <NO_FUNC 0xf016f1 0xf016f1 NO_FUNC
+					 NO_FUNC 0xf016f5 0xf016f5 NO_FUNC>;
+			func3_en_mask = <0       0x20     0x20     0
+					 0       0x04     0x08     0      >;
```
GPIOH5 and GPIOH6.  Also need to set U2CTRL>0 and UART1PSEL=1.

GTLin08 replied:
U2CTRL>0 has been defined in soc.c(IT8XXX2_GPIO_GRC1 |= BIT(2)).
In function of pinmux_it8xxx2_init:
TODO: If UART2 swaps from bit2:1 to bit6:5 in H group, we have to set UART1PSEL = 1 in UART1PMR register.

keith-zephyr commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -42,9 +43,174 @@
+			func3_gcr =     <NO_FUNC NO_FUNC  NO_FUNC  NO_FUNC
+					 NO_FUNC 0xf016f0 0xf016f0 0xf016f0>;
+			func3_en_mask = <0       0        0        0
+					 0       0x04     0x04     0x04    >;
```
GPIOI5-6: U2CTRL>1, so I think the mask needs to be 0x08.

GTLin08 replied:
Yeah, it's wrong. Corrected.

keith-zephyr commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,159 @@
-	*func = (val & (IT8XXX2_PINMUX_IOF1 << pin)) ?
-		IT8XXX2_PINMUX_IOF1 : IT8XXX2_PINMUX_IOF0;
+
+	*func = (*reg_gpcr & (GPCR_PORT_PIN_MODE_INPUT |
+		GPCR_PORT_PIN_MODE_OUTPUT)) ?
+		IT8XXX2_PINMUX_FUNC_2 : IT8XXX2_PINMUX_FUNC_1;
```
Is the current alternate function set to 2, only if MODE_INPUT is set and MODE_OUTPUT is clear?  This sets *func = FUNC_2 if only MODE_OUTPUT is set.

GTLin08 replied:
Done

keith-zephyr commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,159 @@
+	volatile uint8_t *reg_gpcr =
+		(uint8_t *)(pinmux_config->reg_gpcr + pin);
+
+	if (func == PINMUX_INPUT_ENABLED) {
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_INPUT) &
+			~GPCR_PORT_PIN_MODE_OUTPUT;
+	} else { /* func == PINMUX_OUTPUT_ENABLED */
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_OUTPUT) &
+			~GPCR_PORT_PIN_MODE_INPUT;
+	}
```
Change to verify func==OUTPUT, instead of making it the default.  I also suggest reading the gpcr register once.

gpcr = *reg_gpcr & ~(MODE_INPUT | MODE_OUTPUT);
if (func == INPUT)
    *reg_gpcr = gpcr | MODE_INPUT;
else if (func == OUTPUT)
    *reg_gpcr = gpcr | MODE_OUTPUT:
else
    return -EINVAL;

GTLin08 replied:
Done

keith-zephyr commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,159 @@
+		*reg_gpcr &= ~(GPCR_PORT_PIN_MODE_PULLUP |
+			GPCR_PORT_PIN_MODE_PULLDOWN);
+	}
+
```
Check for func==PULLUP_DISABLE explicitly.
if (func == PULLUP_ENABLE)
else if (func == PULLUP_DISABLE)
else
   return -EINVAL;

GTLin08 replied:
Done

GTLin08 committed:
ite: drivers/pinmux: modify pinmux driver

Modify the pinmux control method and add support the fun3
& fun4 alternation function.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,159 @@
+	case IT8XXX2_PINMUX_FUNC_2:
+		/* Func2: WUI function: turn the pin into an input */
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_INPUT) &
+			~GPCR_PORT_PIN_MODE_OUTPUT;
```
Done

keith-zephyr commented on drivers/pinmux/pinmux_ite_it8xxx2.c:
```diff
@@ -13,76 +13,163 @@
-		IT8XXX2_PINMUX_IOF1 : IT8XXX2_PINMUX_IOF0;
+
+	*func = ((*reg_gpcr & (GPCR_PORT_PIN_MODE_INPUT |
+		GPCR_PORT_PIN_MODE_OUTPUT)) >> 6) == 0x02 ?
```
Nit - this is clearer as:

*func = (*reg_gpcr & (GPCR_PORT_PIN_MODE_INPUT |
		GPCR_PORT_PIN_MODE_OUTPUT)) == GPCR_PORT_PIN_MODE_INPUT);

GTLin08 replied:
Done
